cmd/crypto.go

```diff
@@ -0,0 +1,54 @@
+// Copyright 2016-2019, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package cmd
+
+import (
+	"reflect"
+
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi/pkg/backend"
+	"github.com/pulumi/pulumi/pkg/backend/filestate"
+	"github.com/pulumi/pulumi/pkg/backend/httpstate"
+	"github.com/pulumi/pulumi/pkg/resource/config"
+	"github.com/pulumi/pulumi/pkg/secrets"
+)
+
+func getStackEncrypter(s backend.Stack) (config.Encrypter, error) {
+	sm, err := getStackSecretsManager(s)
+	if err != nil {
+		return nil, err
+	}
+
+	return sm.Encrypter()
+}
+
+func getStackDencrypter(s backend.Stack) (config.Decrypter, error) {
+	sm, err := getStackSecretsManager(s)
+	if err != nil {
+		return nil, err
+	}
+
+	return sm.Decrypter()
+}
+
+func getStackSecretsManager(s backend.Stack) (secrets.Manager, error) {
```

Comment:
Why not just make this a method on the Stack? E.g. stack.SecretsManager(). If you did that it seems a lot of the code would be cleaner, with fewer casts, and nicely isolated in the backend code instead of sprinkled into cmd (which I am hoping over time loses more and more machinery, for many reasons, including ISV partners in hosted scenarios will need to replicate all of it).

Reply:
I'm going to try this.  The reason I didn't do this at first was that it forces the implementation of these stacks to know about the existence of the `Pulumi.<stack-name>.yaml` file, which I don't super love.  I do wish long term these can become separable (but maybe the best path forward for that long term is to pull that stuff behind an interface that is provided to stack and can be used to interact with the project model in an abstract way that is not tied to a storage format).